### PR TITLE
Add a new Coprocessor - ViewTTLAware Coprocessor

### PR DESCRIPTION
**Pull request by jpisaac**
* Add a New coprocessor - ViewTTLAware Coprocessor that will intercept scan/get requests to inject a new ViewTTLAware scanner.

The scanner will -
  * Use the row timestamp of the empty column to determine whether row TTL has expired  and mask the rows from underlying query results.
  * Use the row timestamp to delete expired rows when DELETE_VIEW_TTL_EXPIRED flag is present.

### CODE REVIEW

**Comment by jpisaac** [comment]
@ChinmaySKulkarni @kadirozde @yanxinyi

File: phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java [hunk 1]
```diff
@@ -50,10 +50,12 @@
 import org.apache.phoenix.iterate.RegionScannerFactory;
 import org.apache.phoenix.schema.StaleRegionBoundaryCacheException;
 import org.apache.phoenix.schema.types.PUnsignedTinyint;
+import org.apache.phoenix.util.EnvironmentEdgeManager;
 import org.apache.phoenix.util.ScanUtil;
 import org.apache.phoenix.util.ServerUtil;
 import org.apache.phoenix.util.TransactionUtil;
-
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
```
Comment (yanxinyi) [inline phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java]: unused import? I didn't see your changes using LOGGER

**Comment by ChinmaySKulkarni** [comment]
As discussed offline, please rebase your changes so that `ViewTTLIT` doesn't show up as a new file.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/ViewTTLIT.java [hunk 2]
```diff
@@ -394,4 +420,804 @@ public void testViewTTLWithAlterView() throws Exception {
...
+            scan.setAttribute(BaseScannerRegionObserver.EMPTY_COLUMN_QUALIFIER_NAME, emptyColumnName);
+            scan.setAttribute(BaseScannerRegionObserver.DELETE_VIEW_TTL_EXPIRED, PDataType.TRUE_BYTES);
+            scan.setAttribute(BaseScannerRegionObserver.MASK_VIEW_TTL_EXPIRED, PDataType.FALSE_BYTES);
+            scan.setAttribute(BaseScannerRegionObserver.VIEW_TTL, Bytes.toBytes(Long.valueOf(table.getViewTTL())));
+            PhoenixResultSet rs = pstmt.newResultSet(queryPlan.iterator(), queryPlan.getProjector(), queryPlan.getContext());
+            if (rs.next()) {
+                LOGGER.debug(String.format("Number of deleted rows :- " + rs.getString(1)));
+            }
+        }
+    }
```
Comment (ChinmaySKulkarni) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/ViewTTLIT.java]: Skimmed this test and it looks really exhaustive and thorough, awesome! Can we add a test for the cases of inheritance of the VIEW_TTL property for child views of views?

File: phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java [hunk 3]
```diff
@@ -374,20 +380,23 @@ RegionScanner getWrappedScanner(final ObserverContext<RegionCoprocessorEnvironme
                 dataRegion, indexMaintainer, null, viewConstants, null, null, projector, ptr, useQualiferAsListIndex);
     }
 
-    @Override
-    public KeyValueScanner preStoreScannerOpen(final ObserverContext<RegionCoprocessorEnvironment> c,
-        final Store store, final Scan scan, final NavigableSet<byte[]> targetCols,
-        final KeyValueScanner s) throws IOException {
+    @Override public KeyValueScanner preStoreScannerOpen(
```
Comment (ChinmaySKulkarni) [inline phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java]: not sure what the diff is here..seems like mainly whitespace changes, right? Can we revert this?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/ViewTTLIT.java [hunk 4]
```diff
@@ -394,4 +420,804 @@ public void testViewTTLWithAlterView() throws Exception {
...
+                String
+                        col6 =
+                        String.format("f%05d", rowIndex + rnd.nextInt(MAX_ROWS));
+                String
+                        col7 =
+                        String.format("g%05d", rowIndex + rnd.nextInt(MAX_ROWS));
+                String
+                        col8 =
+                        String.format("h%05d", rowIndex + rnd.nextInt(MAX_ROWS));
+                String
```
[the remark this answers is not in the document]
Comment (jpisaac) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/ViewTTLIT.java]: It is formatted using the Phoenix Code template. Do you think we should override that? We will be struggling with manually formatting everytime.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/ViewTTLIT.java [hunk 5]
```diff
@@ -394,4 +420,804 @@ public void testViewTTLWithAlterView() throws Exception {
...
+                            QueryConstants.EMPTY_COLUMN_BYTES :
+                            table.getEncodingScheme().encode(QueryConstants.ENCODED_EMPTY_COLUMN_NAME);
+
+            scan.setAttribute(BaseScannerRegionObserver.EMPTY_COLUMN_FAMILY_NAME, emptyColumnFamilyName);
+            scan.setAttribute(BaseScannerRegionObserver.EMPTY_COLUMN_QUALIFIER_NAME, emptyColumnName);
+            scan.setAttribute(BaseScannerRegionObserver.DELETE_VIEW_TTL_EXPIRED, PDataType.TRUE_BYTES);
+            scan.setAttribute(BaseScannerRegionObserver.MASK_VIEW_TTL_EXPIRED, PDataType.FALSE_BYTES);
+            scan.setAttribute(BaseScannerRegionObserver.VIEW_TTL, Bytes.toBytes(Long.valueOf(table.getViewTTL())));
+            PhoenixResultSet rs = pstmt.newResultSet(queryPlan.iterator(), queryPlan.getProjector(), queryPlan.getContext());
+            if (rs.next()) {
```
[the remark this answers is not in the document]
Comment (jpisaac) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/ViewTTLIT.java]: Will remove it, missed that as part of the cleanup.

File: phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java (hunk 3 above)
Comment (jpisaac) [reply phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java]: Again, this is Phoenix code template formatting kicking in.

File: phoenix-core/src/main/java/org/apache/phoenix/coprocessor/TTLAwareRegionObserver.java [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+
+        private boolean isEmptyColumn(Cell cell) {
+            return Bytes.compareTo(cell.getFamilyArray(), cell.getFamilyOffset(), cell.getFamilyLength(),
+                    emptyCF, 0, emptyCF.length) == 0 &&
+                    Bytes.compareTo(cell.getQualifierArray(), cell.getQualifierOffset(), cell.getQualifierLength(),
+                            emptyCQ, 0, emptyCQ.length) == 0;
+        }
+
+        // TODO : Remove it after we verify all SQLs include the empty column.
+        private boolean checkEmptyColumnNotExpired(byte[] rowKey) throws IOException {
```
Comment (yanxinyi) [inline phoenix-core/src/main/java/org/apache/phoenix/coprocessor/TTLAwareRegionObserver.java]: can you explain more here and why the current region doesn't have an empty column?

File: phoenix-core/src/main/java/org/apache/phoenix/util/IndexUtil.java [hunk 7]
```diff
@@ -39,11 +37,12 @@
 import java.util.List;
 import java.util.ListIterator;
 import java.util.Map;
-import java.util.NavigableSet;
 import java.util.concurrent.TimeUnit;
 
 import com.google.common.cache.Cache;
 import com.google.common.cache.CacheBuilder;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
```
Comment (yanxinyi) [inline phoenix-core/src/main/java/org/apache/phoenix/util/IndexUtil.java]: nit: new import statement but no use in this file?

File: phoenix-core/src/main/java/org/apache/phoenix/util/ScanUtil.java [hunk 8]
```diff
@@ -115,16 +133,37 @@ public static boolean isLocalIndex(Scan scan) {
...
+    public static long getViewTTL(Scan scan) {
+        byte[] viewTTL = scan.getAttribute(BaseScannerRegionObserver.VIEW_TTL);
+        if (viewTTL == null) {
+            return 0L;
+        }
+        return Bytes.toLong(viewTTL);
+    }
+
+    public static boolean isMaskTTLExpiredRows(Scan scan) {
+        return scan.getAttribute(BaseScannerRegionObserver.MASK_VIEW_TTL_EXPIRED) != null && (Bytes.compareTo(scan.getAttribute(BaseScannerRegionObserver.MASK_VIEW_TTL_EXPIRED),
```
Comment (yanxinyi) [inline phoenix-core/src/main/java/org/apache/phoenix/util/ScanUtil.java]: nit: more than 100 chars per line.

File: phoenix-core/src/main/java/org/apache/phoenix/query/ConnectionQueryServicesImpl.java [hunk 9]
```diff
@@ -1066,6 +1067,13 @@ private void addCoprocessors(byte[] tableName, HTableDescriptor descriptor, PTab
                     }
                 }
             }
+
+            if (!SchemaUtil.isSystemTable(tableName)) {
+                if (!descriptor.hasCoprocessor(TTLAwareRegionObserver.class.getName())) {
+                    descriptor.addCoprocessor(TTLAwareRegionObserver.class.getName(), null, priority-2, null);
```
Comment (ChinmaySKulkarni) [inline phoenix-core/src/main/java/org/apache/phoenix/query/ConnectionQueryServicesImpl.java]: Can you please add a comment to mention why this Co-processor priority has to be so high?

File: phoenix-core/src/main/java/org/apache/phoenix/util/IndexUtil.java (hunk 7 above)
Comment (ChinmaySKulkarni) [reply phoenix-core/src/main/java/org/apache/phoenix/util/IndexUtil.java]: +1

File: phoenix-core/src/main/java/org/apache/phoenix/util/ScanUtil.java [hunk 10]
```diff
@@ -956,8 +1026,130 @@ public static int getClientVersion(Scan scan) {
         }
         return clientVersion;
     }
-    
+
     public static void setClientVersion(Scan scan, int version) {
         scan.setAttribute(BaseScannerRegionObserver.CLIENT_VERSION, Bytes.toBytes(version));
     }
+
```
Comment (ChinmaySKulkarni) [inline phoenix-core/src/main/java/org/apache/phoenix/util/ScanUtil.java]: Most of the diffs in ScanUtil are formatting changes. Can we revert them?

Comment (yanxinyi) [reply phoenix-core/src/main/java/org/apache/phoenix/util/ScanUtil.java]: I think it has to do with IDE auto formating

File: phoenix-core/src/main/java/org/apache/phoenix/coprocessor/TTLAwareRegionObserver.java [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+            boolean isRowExpired = !checkRowNotExpired(cellList);
+            if (isRowExpired) {
+                long ttl = ScanUtil.getViewTTL(this.scan) ;
+                long ts = getMaxTimestamp(cellList);
+                LOG.debug(String.format("***** VIEW-TTL: Deleting region = %s, row = %s, delete-ts = %d, max-ts = %d ****** ",
+                        region.getRegionInfo().getTable().getNameAsString(),
+                        Bytes.toString(rowKey),
+                        now-ttl, ts));
+                Delete del = new Delete(rowKey, now-ttl);
+                //del.addFamily(CellUtil.cloneFamily(firstCell));
```
Comment (yanxinyi) [inline phoenix-core/src/main/java/org/apache/phoenix/coprocessor/TTLAwareRegionObserver.java]: nit: remove?

**Comment by yanxinyi** [comment]
+ 1 
Overall looks great, thanks, @jpisaac  

@ChinmaySKulkarni can you review it again just in case I am missing any critical point here.

File: phoenix-core/src/main/java/org/apache/phoenix/coprocessor/TTLAwareRegionObserver.java (hunk 11 above)
Comment (ChinmaySKulkarni) [reply phoenix-core/src/main/java/org/apache/phoenix/coprocessor/TTLAwareRegionObserver.java]: @jpisaac can you remove this commented code?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/ViewTTLIT.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (ChinmaySKulkarni) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/ViewTTLIT.java]: @jpisaac can you remove the logs?

**Comment by jpisaac** [comment]
@yanxinyi  @ChinmaySKulkarni

**Comment by jpisaac** [comment]
Thanks @ChinmaySKulkarni

**Pull request closed without merging**